agent/consul/state/catalog.go

```diff
@@ -2179,3 +2366,194 @@ func checkSessionsTxn(tx *memdb.Txn, hc *structs.HealthCheck) ([]*sessionCheck,
+
+	// Delete all associated with gateway first, to avoid keeping mappings that were removed
+	if _, err := tx.DeleteAll("terminating-gateway-services", "gateway", structs.NewServiceID(entry.Name, entMeta)); err != nil {
+		return fmt.Errorf("failed to truncate gateway services table: %v", err)
+	}
+
+	gatewayID := structs.NewServiceID(entry.Name, &entry.EnterpriseMeta)
+	for _, svc := range entry.Services {
+		// If the service is a wildcard we need to target all services within the namespace
+		if svc.Name == structs.WildcardSpecifier {
+			services, err := s.catalogServiceListByKind(tx, structs.ServiceKindTypical, entMeta)
+			if err != nil {
+				return fmt.Errorf("failed querying services: %s", err)
+			}
+
+			// Iterate over them and insert
+			for service := services.Next(); service != nil; service = services.Next() {
+				sn := service.(*structs.ServiceNode)
+
+				// Only associate non-consul services with gateways
+				if sn.ServiceName == "consul" {
+					continue
+				}
+
+				existing, err := s.serviceTerminatingGateway(tx, sn.ServiceName, &sn.EnterpriseMeta)
+				if err != nil {
+					return fmt.Errorf("gateway service lookup failed: %s", err)
+				}
+
+				if existing != nil {
+					// Return an error if the wildcard is attempting to cover a service specified by a different gateway's config entry
+					cfg := existing.(*structs.GatewayService)
+					if !cfg.Gateway.Matches(&gatewayID) {
+						return fmt.Errorf("service %q is associated with different gateway, %q", svc.Name, cfg.Gateway)
+					}
+
+					// If there's an existing service associated with this gateway then we skip it.
+					// This means the service was specified on its own, and the service entry overrides the wildcard entry.
+					continue
+				}
+
+				mapping := &structs.GatewayService{
+					Gateway:     gatewayID,
+					Service:     structs.NewServiceID(sn.ServiceName, &svc.EnterpriseMeta),
+					GatewayKind: structs.ServiceKindTerminatingGateway,
+					KeyFile:     svc.KeyFile,
+					CertFile:    svc.CertFile,
+					CAFile:      svc.CAFile,
+				}
+				if err := tx.Insert("terminating-gateway-services", mapping); err != nil {
+					return fmt.Errorf("failed inserting gateway service mapping: %s", err)
+				}
+			}
+
+			// Also store a mapping for the wildcard so that the TLS creds can be pulled
+			// for new services registered in its namespace
+			mapping := &structs.GatewayService{
+				Gateway:     gatewayID,
+				Service:     structs.NewServiceID(svc.Name, &svc.EnterpriseMeta),
+				GatewayKind: structs.ServiceKindTerminatingGateway,
+				KeyFile:     svc.KeyFile,
+				CertFile:    svc.CertFile,
+				CAFile:      svc.CAFile,
+			}
+			if err := tx.Insert("terminating-gateway-services", mapping); err != nil {
+				return fmt.Errorf("failed inserting gateway service mapping: %s", err)
+			}
+
+			continue
+		}
+
+		// Check if the service is already associated with a gateway
+		existing, err := s.serviceTerminatingGateway(tx, svc.Name, &svc.EnterpriseMeta)
```

Comment:
On my first pass at reading this function I thought that there was a bug here with processing the request both as a wildcard and then again outside of that conditional. Then I figured out that the last `continue` a few lines up was causing all this code to not run.

I think this function could be much easier to follow if both cases (wildcard service and singular service) were pulled out into their own functions. That would eliminate the nested for loops with continues at multiple levels which hopefully would be easier to determine at a glance the control flow through the function.

Reply:
I added a function for the wildcard, now that the `continue` is close to the main loop it should be clearer that the code below it is skipped.

Reply:
Yeah, I think the original being less clear was because of the nested loops with their own continues so you had to then track back to figure out which loop you are continuing. This is much better.